### PR TITLE
Support self managed capability

### PR DESCRIPTION
**Pull request by luxiliu**
This is to support self-managed video call based on @danjenkins's changes.

### CODE REVIEW

**Comment by codetheweb** [comment]
I tried this PR but kept getting the below error:

Screen Shot 2020-07-06 at 11 06 19 AM

Do I need to change any methods or setup calls from the current version of this package?

```js
RNCallKeep.setup({
      ios: {
        appName: 'Sample App',
      },
      android: {
        alertTitle: 'Permissions required',
        alertDescription: 'Sample App needs to be able to accept incoming calls.',
        cancelButton: 'Cancel',
        okButton: 'ok',
        imageName: 'phone_account_icon',
        selfManaged: true,
      },
    });
```

**Comment by luxiliu** [comment]
> I tried this PR but kept getting the below error:
> 
> Screen Shot 2020-07-06 at 11 06 19 AM
> 
> Do I need to change any methods or setup calls from the current version of this package?
> 
> ```js
> RNCallKeep.setup({
>       ios: {
>         appName: 'Sample App',
>       },
>       android: {
>         alertTitle: 'Permissions required',
>         alertDescription: 'Sample App needs to be able to accept incoming calls.',
>         cancelButton: 'Cancel',
>         okButton: 'ok',
>         imageName: 'phone_account_icon',
>         selfManaged: true,
>       },
>     });
> ```

Hi @codetheweb , do you still see this error?

**Comment by codetheweb** [comment]
Sorry, I forgot to reply.
We're no longer using this package so I can't currently test it.

**Comment by meliodev** [comment]
Hey @luxiliu ! I have just tried your latest PR, and I've got the same error 'selfManaged' as @codetheweb. Do you have please a fix in mind for this ?

**Comment by meliodev** [comment]
Hey @luxiliu ! I have just tried your latest PR, and I've got the same error 'selfManaged' as @codetheweb. Do you have please a fix in mind for this ?

**Comment by luxiliu** [comment]
Hi @meliodev, sorry, just updated, can you have a try again?

**Comment by meliodev** [comment]
Hi @luxiliu ! 
Here is what I did: 

1. AndroidManifest.xml

```
<uses-permission android:name="android.permission.MANAGE_OWN_CALLS"/>
<uses-permission android:name="android.permission.READ_CALL_LOG"/>
```

2.
```
    android: {
        . . .
        selfManaged: true,  // <- I added this 
      }
```

But nothing happens when I call
`RNCallKeep.displayIncomingCall(callUUID, 'X', 'Y', 'number', true);`

**Comment by luxiliu** [comment]
@meliodev , `RNCallKeep.displayIncomingCall()` is to trigger system native incoming call ui. Here because it's `selfManaged`, it won't use the native ui. Yes, there's a new callback. I'll post a snippet later

**Comment by luxiliu** [comment]
@meliodev , I've updated the example, hope it could help you.

File: android/src/main/java/io/wazo/callkeep/VoiceConnectionService.java [hunk 1]
```diff
@@ -248,6 +235,13 @@ private Connection createConnection(ConnectionRequest request) {
         extrasMap.put(EXTRA_CALL_IDENTIFIER, request.getAddress().toString());
         VoiceConnection connection = new VoiceConnection(this, extrasMap);
         connection.setConnectionCapabilities(Connection.CAPABILITY_MUTE);
+
+        TelecomManager telecomManager = (TelecomManager) this.getApplicationContext().getSystemService(Context.TELECOM_SERVICE);
+        PhoneAccount phoneAccount = telecomManager.getPhoneAccount(request.	getAccountHandle());
```
Comment (sboily) [inline android/src/main/java/io/wazo/callkeep/VoiceConnectionService.java:240]: Looks like there is extra space here.

**Comment by sboily** [comment]
Hello @luxiliu, i like your approach to solve this new feature, thank you for this update. What is the next step to merge this PR? Do you think it's correct for resolving the video issue on callkeep with Android? I haven't tested but i read the documentation on the Android SDK and it looks like a good solution. Any objections here about this PR?

**Comment by luxiliu** [comment]
@sboily , next step is probably to implement VideoProvider, which is to use system managed ui, but that could be in another PR.  @danjenkins , do you have any concerns about this PR?

**Comment by meliodev** [comment]
Hi again @luxiliu after updating the example I have got this error 
react-native-callkeep-error 1

**Comment by luxiliu** [comment]
@meloidev are you able to run the example?

**Comment by meliodev** [comment]
@luxiliu I just run the example as it is. 
So when I pressed the first time on 'Display incoming call now', showIncomingCall event was triggered. But no ringtone and pop up for answer/decline call shows up. Honestly, I don't know exactly the calling flow I have to use. Am I supposed to use another library to show an answer/call notification with ringtone ? Can you please clarify to me how I can proceed. Thank you again for your help.

**Comment by luxiliu** [comment]
@meliodev , I'm afraid you need to have a look https://developer.android.com/guide/topics/connectivity/telecom/selfManaged

**Comment by meliodev** [comment]
@luxiliu first thank you for the awesome work you have provided to this library. 
Please I just want some directives as I am stuck with this issue since long time :'( . Can you please tell me briefly the next steps I have to do before redirecting the user to my own videoCall UI. 
Here is the flow I am expecting and please correct me if I am wrong.

1. The callee receives a firebase cloud notification which triggers a ringing UI (even when the app is killed and phone is locked)
2. The callee accepts or rejects the call.
3. The callee is redirected to the video call UI if he accepts the call.

Now my main question is
- How can I build an incomingCall UI with ringtone. Is there an existing react native library for that ?

**Comment by luxiliu** [comment]
@meliodev, sorry, I don't know about such libraries. I'm afraid you have to explore it yourself.

**Comment by sboily** [comment]
Hello, i checked this PR and it add self managed capability to callkeep and not specially concentrated to the video support. I propose to renamed it to reflect this interesting work. If you are interesting about video support on callkeep, i open this PR #251. It add video support on callkeep. So basically it replace the Answer button by the video button etc ... This i not yet finished, but feedback is welcome.

**Comment by sboily** [comment]
Please note there is also a work from @Telzio guy here about self managed capability : https://github.com/Telzio/react-native-callkeep/tree/feature-android-self-managed

**Comment by meliodev** [comment]
Yes I still had hope for somebody to help with that :( I really need that feature. I would love help and contribute but unfortunately I don't have enough knowledge on that.

**Comment by luxiliu** [comment]
I'll create a new one

**Comment by luxiliu** [comment]
I was hoping I could polish the code and raise another PR with some video support, but it turned out I couldn't find myself enough time to do the coding, so I think it's better to reopen it. Please provide feedback and re-review it, thanks.

**Comment by luxiliu** [comment]
A new pull request is raised to just focus self-managed call support, https://github.com/react-native-webrtc/react-native-callkeep/pull/310

**Pull request closed without merging**